### PR TITLE
Docker Compose setup and documentation

### PR DESCRIPTION
**Pull request by darinkrauss**
Initial new `docker-compose.yml` and documentation. The big differences are: 1) configuration is injected into the container (rather than being included in the image); and 2) configurability for various forms of development (within and outside of Docker). The default values for all environment variables found in `docker-compose.yml` are found in the `.env` file. This is SOP for Docker. You can either edit the `.env` file to override for your local configuration (but do *NOT* commit your changes) or just define normal environment variables via your shell (login script or `direnv`, as I do). Normal environment variables take precedence over those found in the `.env` file, so IMO this is the best solution. I've explicit name-spaced the environment variables with a `TIDEPOOL_DOCKER_` prefix so there should be no collisions.

I highly recommend reading the documentation for details.

The `docker-compose.yml` temporarily (for this PR until merged into `master`) uses specially built and tagged versions of the various images using the `:docker-build` suffix. This will be removed when this and all other `Dockerfile` updates in various repos are merged into `master`.

Pinging everyone in case you are interested. Please try it out and let me know if you have any problems.

Feedback is appreciated. If you feel I've gone in the wrong direction, please speak up. There is still some ways to get to a fully containerized deployment with something like Kubernetes (dealing with ports among several other things). Also, the `hydrophone`, `shoreline,` and `tide-whisperer` configurations "files" are stop-gap measures to be able to inject the configuration without requiring extensive changes to the code (that will eventually be deprecated).

@cbwebdevelopment If you want to move over the `tidepool` script, you can. However, I'm not a huge fan of custom scripts hiding relatively simple base commands. Most of the commands in `tidepool` are simple wrappers around the normal `docker-compose` commands and do little else. The override file isn't necessary given above. OTOH, the `link` and `unlink` commands are complicated enough to warrant a script file. If you do add a script after I merge into `master`, please add it to a `bin` directory to highlight the fact it is executable and to allow the option of adding that directory (rather than the root directory) to the PATH environment variable.

@pazaan @jh-bate @krystophv @cbwebdevelopment @gniezen

### CODE REVIEW

File: docker-compose.yml [hunk 1]
```diff
@@ -0,0 +1,526 @@
...
+      - '27017:27017'
+
+  blip:
+    image: tidepool/blip:docker-build
+    depends_on:
+      - hakken
+    # build: ${TIDEPOOL_DOCKER_BLIP_DIR}
+    # volumes:
+    #   - ${TIDEPOOL_DOCKER_BLIP_DIR}:/app:cached
+    #   - /app/node_modules
```
Comment (pazaan) [inline docker-compose.yml]: What do we need to mount /app/node_modules for? Changes are that it's going to be cross-platform, which could result in unexpected results.

Comment (cbwebdevelopment) [reply docker-compose.yml]: This allows us to mount the application code from our local machines for development purposes, while excluding the `node_modules` directory.

When multiple `node_modules` directories are volume-mounted into docker-for-mac, I've found that the file system events stop working properly (local updates no longer trigger changes in the containers, for instance).

The other issue this seems to help with quite a bit was that due to the watches on the massive amount of files in the `node_modules` directory, CPU and memory usage on my machine was through the roof, and the fan ran continuously before excluding this directory.

These problems do happen when mounting just one service, but are compounded when mounting and linking, for instance, `blip`, `tideline`, `platform-client`, and `viz` at the same time.

As far as cross-platform concerns go, I feel that excluding the `node_modules` directory is a good thing, as it ensures that the npm packages running in the containers are not those coming from the host OS (i.e. OSX or Windows).

File: README.md [hunk 2]
```diff
@@ -0,0 +1,539 @@
...
+
+You can now quit the Docker application, if you wish.
+
+# Customization
+
+## Configuration Secrets, Hash Salts, and TLS Certificate
+
+Your local Tidepool includes a number of environment variables that specify configuration secrets, hash salts, and a self-signed TLS certificate that can be customized to uniquely secure your installation. The default values for these environment variables were copied from the previous local development setup in order to maintain backwards compatibility. However, if your local Tidepool is to be used for anything more than development purposes, you should change these environment variables for security reasons.
+
+NOTE: Changing some of these environment variables **will** cause existing data in your Mongo database to become useable. It is highly recommended that you change the above environment variables *only* if you are willing to create a new Mongo database, thus losing all previously created accounts and uploaded data.
```
Comment (cbwebdevelopment) [inline README.md]: ...cause existing data in your Mongo database to become **useable**

Should this be **unusable**?

**Comment by cbwebdevelopment** [comment]
@darinkrauss Regarding the tidepool helper script, I have mixed feelings about wrapping relatively simple base commands.  I will likely port the script over at some point (after this is merged to master), and may take your cue to remove some of the more simple commands.

The thing I like about them is being able to run the commands from any directory (i.e. not having to run things from the same directory that my `docker-compose.yml` file sits in, or with a flag pointing to it).  Definitely a very minor win, though.

As far as the compose overrides (with `docker-compose.override.yml`) go, I do think it's worth considering leaving that as an option, and perhaps even encouraging this approach in the README.  Like defining environment variables in a `.env` file, this too seems to me to be SOP for Docker (see https://docs.docker.com/compose/extends/).  Docker will read this file by default, and it makes it cleaner for users to freely make modifications as desired to the compose file services without touching the main `docker-compose.yml` file at all, which makes for a nicer process when pulling in updates.  I would recommend going this route, and adding the override file to `.gitignore`.

Because Docker reads this file by default, it can definitely be removed from the `tidepool` script.

As for the link and unlinks, yes, they are worth having a script for.  I need to use them regularly as I switch between branches that require different linked packages.

I find the wrapper for the yarn commands against the `NODE_SERVICES` and `BLIP_MOUNTED_NODE_SERVICES` to be very handy time-savers for me as well.  I use them all the time, and believe they'll be useful to others developing in node.

I think I'd like port most of the functionality over (in a `bin` dir as you've suggested) and let the other JS devs play with the Docker stack for a bit before deciding what's useful enough to keep and what can be discarded.  Or... at the very least, have a meeting with them (@krystophv and @gniezen ) to discuss it.

File: README.md (hunk 2 above)
Comment (darinkrauss) [reply README.md]: Yes, yes it should. :)

**Comment by darinkrauss** [comment]
@cbwebdevelopment Okay, that works for me. After this gets merged into `master`, I'll leave it up to you to add the override file to the `.gitignore` and to the documentation (as an alternative to the shell environment variables, though I think the override file is more appropriate to larger changes such as adding/removing containers, etc.). 

The main reason I used environment variables (rather than either directly changing the `.env` file or using an override file) was the simplicity of complex changes. No need to build a full yaml hierarchy, or build multiple services, just to change a port or volume.

File: docker-compose.yml (hunk 1 above)
Comment (pazaan) [reply docker-compose.yml]: Thanks. I was just misunderstanding how it was being excluded. (I thought it was being _included_, so I absolutely agree with all the reasons you have here).

**Pull request merged**